README.md

```diff
@@ -114,6 +110,45 @@ history.listen((location, action) => {
 })
 ```
 
+Post-change listener functions will be not be invoked when a transition is blocked by a pre-change hook, unless the user used the built-in browser back or forward features to initiate the transition. In this case, the listener functions will be called with the same `location` and `action` as the previous invocation.
+
+#### Pre-change hook
+You can perform side effects before history state changes by registering functions with `history.before`
+
+```js
+const unsub = history.before((location, action) => {
+  location.query = queryParseFn(location.search)
+})
+```
+
+Before hooks can also be `async` functions, i.e. return a Promise, to block the state change asynchronously.
+
+```js
+const unsub = history.before(async (location, action) => {
+  await getDataAndCodeForLocation(location)
+})
+```
+
+Returning a truthy value from any before hook functions will cancel the transition.
+
+```js
+const unsub = history.before((location, action) => {
+  return await askUserIfTheyreSureTheyWantToLeave()
```

Comment:
This use case is solved by `history.block`, or at least it's the use case that prompted that API.